crates/bevy_tasks/src/thread_executor.rs

```diff
@@ -0,0 +1,140 @@
+use std::{
+    marker::PhantomData,
+    sync::Arc,
+    thread::{self, ThreadId},
+};
+
+use async_executor::{Executor, Task};
+use futures_lite::Future;
+
+/// An executor that can only be ticked on the thread it was instantiated on. But
+/// can spawn `Send` tasks from other threads.
+///
+/// # Example
+/// ```rust
+/// # use std::sync::{Arc, atomic::{AtomicI32, Ordering}};
+/// use bevy_tasks::ThreadExecutor;
+///
+/// let thread_executor = ThreadExecutor::new();
+/// let count = Arc::new(AtomicI32::new(0));
+///
+/// // create some owned values that can be moved into another thread
+/// let thread_executor_clone = thread_executor.clone();
+/// let count_clone = count.clone();
+/// let thread_spawner = thread_executor.spawner();
+///
+/// std::thread::scope(|scope| {
+///     scope.spawn(|| {
+///         // we cannot get the ticker from another thread
+///         let not_thread_ticker = thread_executor_clone.ticker();
+///         assert!(not_thread_ticker.is_none());
+///         
+///         // but we can spawn tasks from another thread
+///         thread_spawner.spawn(async move {
+///             count_clone.fetch_add(1, Ordering::Relaxed);
+///         }).detach();
+///     });
+/// });
+///
+/// // the tasks do not make progress unless the executor is manually ticked
+/// assert_eq!(count.load(Ordering::Relaxed), 0);
+///
+/// // tick the ticker until task finishes
+/// let thread_ticker = thread_executor.ticker().unwrap();
+/// thread_ticker.try_tick();
+/// assert_eq!(count.load(Ordering::Relaxed), 1);
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct ThreadExecutor {
+    executor: Arc<Executor<'static>>,
+    thread_id: ThreadId,
+}
+
+impl Default for ThreadExecutor {
+    fn default() -> Self {
+        Self {
+            executor: Arc::new(Executor::new()),
+            thread_id: thread::current().id(),
+        }
+    }
+}
+
+impl ThreadExecutor {
+    /// create a new `[ThreadExecutor]`
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Gets the `[ThreadSpawner]` for the thread executor.
+    /// Use this to spawn tasks that run on the thread this was instatiated on.
+    pub fn spawner(&self) -> ThreadSpawner<'static> {
+        ThreadSpawner(self.executor.clone())
+    }
+
+    /// Gets the `[ThreadTicker]` for this executor.
+    /// Use this to tick the executor.
+    /// It only returns the ticker if it's on the thread the executor was created on
+    /// and returns `None` otherwise.
+    pub fn ticker(&self) -> Option<ThreadTicker> {
+        if thread::current().id() == self.thread_id {
+            return Some(ThreadTicker {
+                executor: self.executor.clone(),
+                _marker: PhantomData::default(),
+            });
+        }
+        None
+    }
+}
+
+/// Used to spawn on the [`ThreadExecutor`]
+#[derive(Debug)]
+pub struct ThreadSpawner<'a>(Arc<Executor<'a>>);
```

Comment:
```suggestion
pub struct ThreadSpawner<'a>(&'a Executor<'static>>);
```

Could we get this to work? Seems doable and also avoids the `Arc::clone`.

Reply:
We can't change the lifetime on the Executor to 'static as that is the lifetime of the futures we're allowed to spawn and that is important for how scope functions. Maybe we can pass a reference to the executor around though? I think I tried it and it didn't work, but can't remember why. I'll try it again.

Reply:
Tried to change it to `pub struct ThreadSpawner<'a, 'b>(&'b Executor<'a>)`, but couldn't figure out the knot of lifetimes that causes.

Reply:
Not super mandatory since the clone isn't on the hotpath for anything.